### PR TITLE
Markdown rendering / Normalized Html styling

### PR DESCRIPTION
**Pull request by corydickson**
Addresses issue #386 

I had to update the `rollup` config with a JSON plugin due to a failed installation. Let me know if you are able to reproduce.

Also since we are using `jsx` everywhere I don't think we'll need to sanitize the input.

### CODE REVIEW

File: src/components/RenderedMarkdown/RenderedMarkdown.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import SafeLink from '../Link/SafeLink'
+import Checkbox from '../Input/Checkbox'
+import { TASK_LIST_CLASSNAME } from '../NormalizedHtml/NormalizedHtml'
+import MDReactComponent from 'react-markdown'
```
Comment (bpierre) [inline src/components/RenderedMarkdown/RenderedMarkdown.js]:
```suggestion
import ReactMarkdown from 'react-markdown'
```

Comment (bpierre) [inline src/components/RenderedMarkdown/RenderedMarkdown.js]: Could we also move this line after the `prop-types` import (to group the global imports together).

File: src/components/RenderedMarkdown/RenderedMarkdown.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  children: PropTypes.node,
+}
+
+class RenderedMarkdown extends React.Component {
+  static propTypes = {
+    content: PropTypes.string,
+  }
+
+  static defaultProps = {
+    content: PropTypes.string,
```
Comment (bpierre) [inline src/components/RenderedMarkdown/RenderedMarkdown.js]:
```suggestion
```

File: src/components/RenderedMarkdown/RenderedMarkdown.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+ListItem.propTypes = {
+  checked: PropTypes.bool,
+  children: PropTypes.node,
+}
+
+class RenderedMarkdown extends React.Component {
+  static propTypes = {
+    content: PropTypes.string,
```
Comment (bpierre) [inline src/components/RenderedMarkdown/RenderedMarkdown.js]: We should probably make this required.
```suggestion
    content: PropTypes.string.isRequired,
```

File: src/components/NormalizedHtml/NormalizedHtml.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  h4,
+  h5,
+  h6 {
+    ${font({ weight: 'normal' })}
+    margin-bottom: 16px;
+    margin-top: 24px;
+  }
+  h1,
+  h2 {
+    padding-bottom: 0.3em;
```
Comment (bpierre) [inline src/components/NormalizedHtml/NormalizedHtml.js]: Could we use `px` everywhere for consistency?

File: devbox/apps/MarkdownRenderer.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import React from 'react'
+import styled from 'styled-components'
+import { NormalizedHtml, RenderedMarkdown } from '@aragon/ui'
+
+const MARKDOWN_HTML = "Italicized text is the *cat's meow*."
```
Comment (bpierre) [inline devbox/apps/MarkdownRenderer.js]: Could we make this a full demo of the different styles applied?

File: src/components/RenderedMarkdown/RenderedMarkdown.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import MDReactComponent from 'react-markdown'
+
+const Link = ({ children, ...props }) => (
+  <SafeLink {...props} target="_blank">
+    {children}
+  </SafeLink>
+)
+
+const ListItem = ({ checked, children, ...props }) => {
+  let checkbox = null
```
Comment (bpierre) [inline src/components/RenderedMarkdown/RenderedMarkdown.js]:
```suggestion
  const checkbox = null
```

File: src/components/RenderedMarkdown/RenderedMarkdown.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  if (checked !== null) {
+    checkbox = <Checkbox checked={checked} />
+  }
+
+  return React.createElement(
```
Comment (bpierre) [inline src/components/RenderedMarkdown/RenderedMarkdown.js]: Is there a specific reason for not using JSX here?

File: src/components/RenderedMarkdown/RenderedMarkdown.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    content: PropTypes.string,
+  }
+
+  render() {
+    const { content } = this.props
+
+    return (
+      <MDReactComponent
+        source={content}
+        renderers={{ link: Link, listItem: ListItem }}
```
Comment (bpierre) [inline src/components/RenderedMarkdown/RenderedMarkdown.js]: Could we also pass custom props here? Maybe using a prop name like `reactMarkdownProps` to make it clear that these are passed to `react-markdown` internally:
 
```jsx
<Markdown content="✌" reactMarkdownProps={ /* … */ } />
```

File: src/components/RenderedMarkdown/RenderedMarkdown.js (hunk 6 above)
Comment (bpierre) [reply src/components/RenderedMarkdown/RenderedMarkdown.js]: Ah sorry I just realized that it was set later. Maybe we can do this check directly in the JSX tree?

File: devbox/apps/MarkdownRenderer.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+class App extends React.Component {
+  render() {
+    return (
+      <Main>
+        <NormalizedHtml>
+          <RenderedMarkdown content={MARKDOWN_HTML} />
+        </NormalizedHtml>
```
[the remark this answers is not in the document]
Comment (chadoh) [reply devbox/apps/MarkdownRenderer.js]: I like this idea.

**Comment by gh1dra** [comment]
@bpierre @chadoh Thank you both for your review! Let me know if there is anything I didn't address from the original issue.

Currently, Travis is failing due to the rollup config: https://travis-ci.org/aragon/aragon-ui/jobs/553950501

Here is the new issue: #400

**Comment by gh1dra** [comment]
@bpierre Is there anything I can do to help push this over the line? Are there any updates using the new design system that you'd like to see incorporated?

**Comment by bpierre** [comment]
Hi @gh1dra, sorry about the super long wait. We are going to review the last changes during the next week or so.

> Are there any updates using the new design system that you'd like to see incorporated?

Ideally yes! Would you mind moving these changes on top of `newstyle` rather than `master`?

Some changes that will be needed for 1.0:

- Replace `theme` by `useTheme()`, and make sure all the colors are coming from it (so that it’s ready for the dark theme).
- Replace `font` by `textStyle()`.
- Use `GU` (e.g. `${2 * GU}px` rather than `16px`.
- Replace `SafeLink` by `Link`.

Let me know if you have any question, I am also `bpierre` on https://aragon.chat :-)

**Comment by corydickson** [comment]
@bpierre Sorry for the delay, rollup is still giving me issues on my local machine but seems to be passing your build pipeline. If this looks good I'll squash the prior commits and add @chadoh as an author

**Comment by bpierre** [comment]
Hey @corydickson, thanks for this, and sorry for the delay!

I added a few commits:

- Move to markdown-to-jsx, which is lighter than react-markdown (5kb vs. 20kb), and doesn’t have the
bundling issues that react-markdown has.
- Fix the theme colors.
- Use `<Link />` and `<Checkbox />` in the rendered markdown.
- Tweak `NormalizedHtml` styles.
- Add documentation.
- Other tweaks, coding style, prop types etc.

I think it is ready to be merged soon :-)

**Comment by corydickson** [comment]
@bpierre Thanks! Do you want to handle to rebase then? Want to be sure @chadoh gets authorship / green squares :)

**Comment by bpierre** [comment]
@corydickson I only merged `master` in, using merge commits! For the autorship issue, did you want to squash your previous commits to make @chadoh the author? If that’s the case feel free to do so, I’ll wait for it before merging. Note that we always squash & merge our PRs, so all the commits here will appear as a single commit on `master`. Let me know!

**Comment by corydickson** [comment]
So when I try to rebase it makes me the author of all the child commits from 728553e onward... let's just add: `Co-authored-by: Chad Ostrowski <hi@chadoh.com>` at the end when you squash after review sorry for the confusion 😅

File: src/components/NormalizedHtml/README.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
+# NormalizedHtml
+
+Apply a default style to its content.
```
[the remark this answers is not in the document]
Comment (bpierre) [reply src/components/NormalizedHtml/README.md:3]: Cool yes let’s keep it internal until we need it :+1:

**Pull request merged**